File: backend/editor/server/flow_assets.py
# -*- coding: utf-8 -*-
"""剧情图媒体资产判定：从全量 AA key 中筛出「CG 图片」与「音乐（BGM）」。

纯函数、不依赖 api.STATE，可脱离 HTTP 服务单测。信号来源：
- 纹理尺寸（aa_index v3 texmeta / 资源包图像头部），CG 为全屏演出图，
  分辨率不低于游戏内 1280x720 是硬底线；
- Addressable 分组编码在 bundle 文件名里（textures_assets_cg / bg / role /
  icon…），作为直通/否决的辅助信号；
- 音频按 audios_assets_bgm bundle 与 AudioCfg(type=1) 的 url 判定 BGM，
  TTS 配音（audio/tts/）与角色语音包（audios_assets_role）排除。
"""
import os
import re

# CG 分辨率硬底线（游戏原生 1280x720）
CG_MIN_W = 1280
CG_MIN_H = 720
# 宽高比合理区间：排除竖幅立绘大图与长条拼接图
CG_RATIO_MIN = 1.15
CG_RATIO_MAX = 2.4

# 确定性剧情演出组（textures_assets_<group>）：分辨率达标即 CG
_CG_GROUPS = frozenset(("cg", "cg2", "big", "comic"))
# 混合内容组（主包/背景/DLC v1xx）：分辨率 + 宽高比启发
_MIXED_GROUPS_PREFIX = ("v",)
_MIXED_GROUPS = frozenset(("", "bg"))
# SpriteAtlas 自动拼页纹理名前缀（icon/role/语音等图集页，非 CG）
_ATLAS_KEY_PREFIX = "sactx-"
_TEXTURES_BUNDLE = "textures_assets_"
_HASH_RE = re.compile(r"[0-9a-f]{16,}$")


def _tex_group(bundle_path):
    """`textures_assets_<group>_<hash>.bundle` → group；非贴图 bundle 返回 None。

    实测 Addressable 打包按组分 bundle（cg/cg2/big/comic/paint/icon/role/
    kzone-head/v177…），组名是最可靠的来源信号；localization、prefabs、
    kzone_min、common 等非 `textures_assets_` 前缀的 bundle 一律 None。
    """
    name = _bundle_name(bundle_path)
    if not name.startswith(_TEXTURES_BUNDLE):
        return None
    rest = name[len(_TEXTURES_BUNDLE):]
    if rest.endswith(".bundle"):
        rest = rest[:-len(".bundle")]
    parts = rest.split("_")
    # 末段是打包 hash（长十六进制）；组名本身可含 '-'（kzone-head）
    if len(parts) > 1 and _HASH_RE.search(parts[-1]):
        parts = parts[:-1]
    return parts[0] if parts else None

# 确定性非音乐：角色语音、配音落盘目录
_MUSIC_REJECT = ("audios_assets_role", "audios_assets_ogg", "audio/tts",
                 "audio\\tts")


def _bundle_name(bundle_path):
    """bundle 路径 → 小写文件名（资源包的解码文件路径同样适用其目录段）。"""
    s = (bundle_path or "").replace("\\", "/").lower()
    return os.path.basename(s) if "/" in s or s else s


def is_cg_image(key, bundle_path, width, height):
    """CG 图片判定：分辨率硬底线 + Addressable 组白名单 + 混合组宽高比启发。

    规则（依次短路）：
    1. w>=1280 且 h>=720（低于底线一律不是 CG）；
    2. `sactx-*`（SpriteAtlas 拼页）否决；
    3. 非 `textures_assets_` 前缀的 bundle（kzone_min/localization/main/
       common/prefabs…）否决；
    4. 确定性演出组 cg/cg2/big/comic 直通；
    5. 混合内容组（主包、bg、DLC v1xx）与资源包平铺文件：宽高比
       1.15~2.4 + 键名启发否决（icon/head/ui/atlas…）。

    width/height 任一缺失（旧索引无 texmeta）返回 False，由调用方决定回退。
    """
    try:
        w = int(width or 0)
        h = int(height or 0)
    except (TypeError, ValueError):
        return False
    if w < CG_MIN_W or h < CG_MIN_H:
        return False
    if (key or "").strip().lower().startswith(_ATLAS_KEY_PREFIX):
        return False
    name = _bundle_name(bundle_path)
    group = _tex_group(name)
    if group is None:
        # 内置资源包平铺文件（pack/tex/*.webp）没有分组信息：走启发式；
        # 其余非贴图 bundle 整包否决
        if name.endswith(".bundle"):
            return False
    elif group in _CG_GROUPS:
        return True
    elif not (group in _MIXED_GROUPS or group.startswith(_MIXED_GROUPS_PREFIX)):
        # paint/icon/role/kzone-head/guide/puzzle/track… 明确非 CG 组
        return False
    ratio = w / float(h)
    if not (CG_RATIO_MIN <= ratio <= CG_RATIO_MAX):
        return False
    kl = (key or "").lower()
    if re.search(r"(icon|head|avatar|cloth|\bui[_-]|atlas|spine)", kl):
        return False
    return True


def music_url_basenames(audio_rows):
    """从 AudioCfg 行（dict-of-dict 或行迭代器）提取 type==1（BGM）的 url basename 集合。"""
    out = set()
    rows = audio_rows.values() if isinstance(audio_rows, dict) else audio_rows
    for row in rows:
        if not isinstance(row, dict):
            continue
        try:
            if int(row.get("type") or 0) != 1:
                continue
        except (TypeError, ValueError):
            continue
        url = str(row.get("url") or "").strip()
        if not url:
            continue
        base = os.path.basename(url.replace("\\", "/")).lower()
        base = os.path.splitext(base)[0]
        if base:
            out.add(base)
    return out


def is_music(key, bundle_path, music_urls):
    """音乐（BGM）判定：bgm bundle 直通 / AudioCfg type=1 url 命中；语音与配音排除。"""
    kl = (key or "").strip().lower()
    name = _bundle_name(bundle_path)
    full = (bundle_path or "").replace("\\", "/").lower()
    for tag in _MUSIC_REJECT:
        if tag in name or tag in full or kl.startswith("audio/tts"):
            return False
    if "audios_assets_bgm" in name or "bgm" in name:
        return True
    return kl in music_urls

File: native/server/services/flow_assets.cpp
// wip/P4/flow_assets.cpp — see flow_assets.h (port of flow_assets.py).
#include "flow_assets.h"

#include <optional>
#include <regex>
#include <string>
#include <vector>

#include "sa_core/strings.h"
#include "sa_core/util.h"
#include "p4_util.h"

namespace sa {
namespace flow {
namespace {

constexpr int kCgMinW = 1280;
constexpr int kCgMinH = 720;
constexpr double kCgRatioMin = 1.15;
constexpr double kCgRatioMax = 2.4;

std::string bundle_name(std::string_view bundle_path) {
    // flow_assets._bundle_name: forward-slash, lowercase, basename.
    std::string s = sa_core::str::lower(sa_core::str::replace_all(std::string(bundle_path), "\\", "/"));
    return p4::basename(s);
}

// flow_assets._tex_group: textures_assets_<group>_<hash>.bundle -> group.
std::optional<std::string> tex_group(const std::string& name) {
    static const std::string kPrefix = "textures_assets_";
    if (!sa_core::str::starts_with(name, kPrefix)) return std::nullopt;
    std::string rest = name.substr(kPrefix.size());
    if (sa_core::str::ends_with(rest, ".bundle"))
        rest = rest.substr(0, rest.size() - std::string(".bundle").size());
    std::vector<std::string> parts;
    size_t pos = 0;
    while (true) {
        size_t under = rest.find('_', pos);
        parts.push_back(rest.substr(pos, under == std::string::npos ? std::string::npos : under - pos));
        if (under == std::string::npos) break;
        pos = under + 1;
    }
    // Trailing 16+ hex segment is the packing hash — drop it.
    static const std::regex kHash("[0-9a-f]{16,}$");
    if (parts.size() > 1 && std::regex_search(parts.back(), kHash)) parts.pop_back();
    if (parts.empty()) return std::nullopt;
    return parts.front();
}

}  // namespace

bool is_cg_image(const std::string& key, const std::string& bundle_path, long long width,
                 long long height) {
    long long w = width;
    long long h = height;
    if (w < kCgMinW || h < kCgMinH) return false;
    std::string kl_lower = sa_core::str::lower(p4::strip(key));
    if (sa_core::str::starts_with(kl_lower, "sactx-")) return false;

    std::string name = bundle_name(bundle_path);
    auto group = tex_group(name);
    if (!group) {
        if (sa_core::str::ends_with(name, ".bundle")) return false;
    } else if (*group == "cg" || *group == "cg2" || *group == "big" || *group == "comic") {
        return true;
    } else if (!(*group == "" || *group == "bg" || sa_core::str::starts_with(*group, "v"))) {
        return false;  // paint/icon/role/kzone-head/guide/...
    }
    double ratio = static_cast<double>(w) / static_cast<double>(h);
    if (!(ratio >= kCgRatioMin && ratio <= kCgRatioMax)) return false;
    static const std::regex kVeto("(icon|head|avatar|cloth|\\bui[_-]|atlas|spine)");
    if (std::regex_search(kl_lower, kVeto)) return false;
    return true;
}

std::set<std::string> music_url_basenames(const nlohmann::ordered_json& audio_rows) {
    std::set<std::string> out;
    // dict-of-rows (values) or an array of rows.
    auto consider = [&](const nlohmann::ordered_json& row) {
        if (!row.is_object()) return;
        long long type = 0;
        if (row.contains("type") && !row.at("type").is_null()) {
            auto iv = p4::json_int(row.at("type"));
            if (!iv) return;  // int() raised -> continue
            type = *iv;
        }
        if (type != 1) return;
        std::string url;
        if (row.contains("url") && !row.at("url").is_null()) url = p4::strip(sa_core::py_str(row.at("url")));
        if (url.empty()) return;
        std::string base = sa_core::str::lower(p4::basename(sa_core::str::replace_all(url, "\\", "/")));
        base = p4::split_ext(base).first;
        if (!base.empty()) out.insert(base);
    };
    if (audio_rows.is_object()) {
        for (auto it = audio_rows.begin(); it != audio_rows.end(); ++it) consider(it.value());
    } else if (audio_rows.is_array()) {
        for (const auto& row : audio_rows) consider(row);
    }
    return out;
}

bool is_music(const std::string& key, const std::string& bundle_path,
              const std::set<std::string>& music_urls) {
    std::string kl = sa_core::str::lower(p4::strip(key));
    std::string name = bundle_name(bundle_path);
    std::string full = sa_core::str::lower(sa_core::str::replace_all(bundle_path, "\\", "/"));
    static const char* kReject[] = {"audios_assets_role", "audios_assets_ogg", "audio/tts",
                                    "audio\\tts"};
    for (const char* tag : kReject) {
        bool in_name = name.find(tag) != std::string::npos;
        bool in_full = full.find(tag) != std::string::npos;
        if (in_name || in_full || sa_core::str::starts_with(kl, "audio/tts")) return false;
    }
    if (name.find("audios_assets_bgm") != std::string::npos ||
        name.find("bgm") != std::string::npos)
        return true;
    return music_urls.count(kl) > 0;
}

}  // namespace flow
}  // namespace sa
